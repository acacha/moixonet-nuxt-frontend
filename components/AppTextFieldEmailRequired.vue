<template>
  <v-text-field
    :value="value"
    :error-messages="errors"
    @input="onInput"
    @blur="$v.value.$touch()"
    :name="name"
    :label="label"
    :data-test="dataTest"
    required
    prepend-icon="mdi-email"
  />
</template>

<script>
import { required, email } from 'vuelidate/lib/validators'

export default {
  name: 'AppTextFieldEmailRequired',
  validations: {
    value: { required, email }
  },
  props: {
    value: {
      type: String,
      required: true
    },
    label: {
      type: String,
      default: 'Correu electrònic'
    },
    name: {
      type: String,
      default: 'email'
    },
    dataTest: {
      type: String,
      default: 'email_field_required'
    }
  },
  data () {
    return {
      email: ''
    }
  },
  computed: {
    errors () {
      const errors = []
      if (!this.$v.value.$dirty) { return errors } // dirty indica si s'ha canviat/tocat el camp
      !this.$v.value.required && errors.push('El camp ' + this.name + ' és obligatori')
      !this.$v.value.email && errors.push('El camp ' + this.name + ' ha de ser un email vàlid')
      return errors
    }
  },
  methods: {
    onInput () {
      this.$v.value.$touch()
      this.$emit('input', arguments[0])
    }
  }
}
</script>
