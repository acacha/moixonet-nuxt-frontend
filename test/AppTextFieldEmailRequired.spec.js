import { mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuelidate from 'vuelidate'
import Vue from 'vue'
import AppTextFieldEmailRequired from '../components/AppTextFieldEmailRequired.vue'

Vue.use(Vuetify)
Vue.use(Vuelidate)

describe('AppTextFieldEmailRequired', () => {
  let opts

  const build = (options) => {
    options = options || opts
    const wrapper = mount(AppTextFieldEmailRequired, options)
    return {
      wrapper,
      inputField: () => wrapper.find('[data-test=email_field_required]')
    }
  }

  beforeEach(() => {
    opts = {
      // sync: false, // avoid inconsistent errors with vuelidate https://github.com/vuelidate/vuelidate/issues/521
      propsData: {
        value: '',
        label: 'Correu electrònic',
        name: 'email'
      }
    }
  })

  test('is a Vue instance', () => {
    const { wrapper } = build()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders the component', () => {
    const { wrapper } = build()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('emits input when typing on input field', async () => {
    const { wrapper, inputField } = build()

    inputField().element.value = 'a'
    inputField().trigger('input')

    await wrapper.vm.$nextTick() // <- over here

    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().input.length).toBe(1)
    expect(wrapper.emitted().input[0][0]).toEqual('a')
  })

  test.skip('show error when field email is and invalid email', async () => {
    const { wrapper, inputField } = build()

    inputField().element.value = 'invalidemailgmail.com'
    inputField().trigger('input')

    await Vue.nextTick()

    console.log(wrapper.text())

    expect(wrapper.text()).toMatch('El camp email ha de ser un email vàlid')
  })

  test.skip('Do not show errors when field email is valid', async () => {
    const { wrapper, inputField } = build()

    inputField().element.value = 'pepe@pardo.com'
    inputField().trigger('input')

    await Vue.nextTick()

    console.log(wrapper.text())
    expect(wrapper.text()).toMatch('El camp email és obligatori')
  })
})
