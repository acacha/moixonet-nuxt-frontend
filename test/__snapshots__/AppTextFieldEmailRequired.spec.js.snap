// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTextFieldEmailRequired renders the component 1`] = `
<div
  class="v-input theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate mdi mdi-email theme--light"
      />
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label theme--light"
          for="input-14"
          style="right: 0px; position: absolute;"
        >
          Correu electr√≤nic
        </label>
        <input
          data-test="email_field_required"
          id="input-14"
          name="email"
          required="required"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;
